Viewmodel phosphor widget implementation

jasongrout opened the pull request:
This implementation follows design discussions about designing the models.

jasongrout committed:
Remove experimental javascript code

jasongrout committed:
Update package.json for the typescript version

jasongrout committed:
Add view model interfaces and reference implementation.

blink1073 commented on src/InputAreaViewModel.ts:
```diff
@@ -0,0 +1,403 @@
+'use strict';
+
+import {
+ISignal, Signal
```
Two leading spaces.

blink1073 commented on src/InputAreaViewModel.ts:
```diff
@@ -0,0 +1,403 @@
+  mimetype: string;
+  
+  /**
+   * Whether or not to display the line numbers in the text editor.
```
I'd remove "or not".

blink1073 commented on src/InputAreaViewModel.ts:
```diff
@@ -0,0 +1,403 @@
+/**
+ * An implementation of an input area view model.
+ */
+export class InputAreaViewModel implements IInputAreaViewModel {
```
`export` should be on a separate line

blink1073 commented on src/InputAreaViewModel.ts:
```diff
@@ -0,0 +1,403 @@
+  Property
+} from 'phosphor-properties';
+
+/**
```
Two blank lines between module-level constructs.

jasongrout commented:
@blink1073 - I cleaned up a lot and I think I fixed each one of your comments.

blink1073 commented on example/index.ts:
```diff
@@ -0,0 +1,27 @@
+  `
+  textModel.mimetype = 'text/x-python'
+  textModel.lineNumbers = true;
+/*  let inputVModel = new InputAreaViewModel();
```
leftover?

jasongrout replied:
Yep....more like, "preview of work to come..."

Commit:
Cleanup

Commit:
More cleanup

blink1073 commented on src/InputAreaViewModel.ts:
```diff
@@ -0,0 +1,405 @@
+    InputAreaViewModel.executionCountProperty.set(this, value);
+  }
+
+    textEditor: ITextEditorViewModel;
```
This should be a getter/setter

jasongrout replied:
Why?

blink1073 replied:
Just an overall preference not to have public attributes.

jasongrout replied:
In general, I'd say that to keep things simple, if you don't need a getter/setter, don't use one.  However, in this case, we need to trigger an update signal so that the widgets know that that whole part of the state changed.  My bad not realizing this at first.  Thanks.

blink1073 commented on src/InputAreaWidget.ts:
```diff
@@ -0,0 +1,188 @@
+  /**
+   * Update the fixedHeight
+   */
+  updateFixedHeight(fixedHeight: boolean) {
```
Why are these not getter/setters?

jasongrout replied:
They are essentially message handlers for model change events.  They aren't state, so much as they update state of the DOM.  They can be easily overridden by subclasses.  Is that enough reasons?  What do you think?

blink1073 replied:
In that case maybe `protected` methods?

blink1073 commented on typings/diff-match-patch/diff-match-patch.d.ts:
```diff
@@ -0,0 +1,15 @@
+// Type definitions for SystemJS 1.6
```
Wrong library.

jasongrout replied:
Right...left overs.

Commit:
Remove extra line

blink1073 commented on src/InputAreaWidget.ts:
```diff
@@ -0,0 +1,194 @@
+   */
+  private _modelUpdate(sender: ITextEditorViewModel, args: IChangedArgs<any>) {
+    switch(args.name) {
+      case 'fixedHeight':
```
Forgot to do this in my PR: `case` should be vertically aligned with `switch`.

jasongrout replied:
I prefer them indented.  In the typescript source, they are all indented too (for example, see https://github.com/Microsoft/TypeScript/blob/331d26f40271c7a21c6fa74ffac633b45f653dda/tests/baselines/reference/switchStatementsWithMultipleDefaults1.js)

blink1073 replied:
Fair enough, this was a Phosphor convention.

jasongrout replied:
I vote we change it :).

Edit: I'm curious to hear @sccolbert's reasoning behind it.  I think logically case statements should be indented because they are inside a block.

sccolbert replied:
Cases are simply labels within the block defined by the switch statement. They are not block-defining statements on their own. So the code within the block defined by the switch is indented as normal.

sccolbert replied:
Some addition points to ponder: http://stackoverflow.com/questions/4299729/why-dont-people-indent-c-access-specifiers-case-statements

jasongrout replied:
being labels, this seems more logical:

``` typescript
switch(c) {
  case 'a':
  f(a)
  break;

  case 'b':
  g(a);
  break;

  case 'c':
  h(a);
  break;
}
```

sccolbert replied:
I'm not going to argue about this either way :-). In Phosphor we align switch and case. I have no opinion for this repo.

jasongrout replied:
all right, I've read up on enough conventions to be convinced it's not crazy :), though there are pretty strong arguments also on the other side.  I'll change my code.

jasongrout committed:
Change the text editor view model to be a property.

jasongrout committed:
Fix comments

jasongrout committed:
Factor out code in the example.

jasongrout committed:
dedent switch case labels

jasongrout committed:
Add a simple implementation of an input area widget.

blink1073 commented:
@jasongrout, are you ready to merge and iterate on this one too?

blink1073 commented on src/InputAreaWidget.ts:
```diff
@@ -0,0 +1,194 @@
+   */
+  constructor(model: ITextEditorViewModel) {
+    super();
+    this.addClass('CodeMirrorWidget');
```
Should be `jp-CodeMirrorWidget`.

jasongrout replied:
I didn't realize we had settled on a prefix.  Sounds good.

jasongrout commented:
@blink1073 - sure we can merge and iterate from here.

jasongrout committed:
Update CSS class names with jp- prefix

blink1073 commented on src/InputAreaWidget.ts:
```diff
@@ -0,0 +1,261 @@
+    this.addClass('InputAreaWidget');
+    this._model = model;
+    for (let key in InputAreaWidget.update) {
+      InputAreaWidget.update[key].call(this, model[key]);
```
This line is causing a build failure.  It should be: `InputAreaWidget.update[key].call(this, (model as any)[key]);`

sccolbert replied:
:-1: on this flavor of dynamic dispatch. It's not type-safe and IMHO is harder to follow.

jasongrout replied:
yep, this was one of the experiments, and I already decided to scrap it.  I have some commits deleting it that I haven't pushed.

jasongrout committed:
Switch back to switch-based message dispatching for input areas.

jasongrout commented:
ping @blink1073 - I switched back to using a switch statement

sccolbert commented on src/InputAreaWidget.ts:
```diff
@@ -0,0 +1,270 @@
+    this._editor.on('change', (instance, change) => {
+      this._model.text = this._editor.getDoc().getValue();
+    });
+    model.stateChanged.connect(this._modelUpdate, this);
```
My convention for naming signal handlers is `onThingSignalName`, so for this case it would be `_onModelStateChanged`. Not saying you should follow this, just letting you know my convention. This let's me look at a method name and immediately make a good guess as to what it's connected to.

jasongrout replied:
Good idea.  Thanks.

sccolbert commented on src/InputAreaWidget.ts:
```diff
@@ -0,0 +1,270 @@
+    if (fixedHeight) {
+      this.addClass('jp-CodeMirrorWidget-fixedHeight');
+    } else {
+      this.removeClass('jp-CodeMirrorWidget-fixedHeight');
```
you can replace this `if/else` block with just `this.toggleClass('jp-CodeMirrorWidget-fixedHeight', fixedHeight)`.

jasongrout replied:
Nice.  Thanks.

sccolbert commented on src/InputAreaWidget.ts:
```diff
@@ -0,0 +1,270 @@
+   * place by using the bitap algorithm to find the corresponding
+   * position of the cursor in the new text.
+   */
+  updateText(text: string) {
```
should all of these `update*` methods be either `protected` or `private`? Is there a strong use case for making them public?

jasongrout replied:
I don't think there is a strong use case.  I'll make them protected.

sccolbert commented on src/InputAreaWidget.ts:
```diff
@@ -0,0 +1,270 @@
+  /**
+   * Get the text in the widget.
+   */
+  get text(): string {
```
should this be public? What about having a read-only getter for the `editor` and calling it done?

jasongrout replied:
I don't want to expose CodeMirror (unless there is a good usecase for making code specific to codemirror), so I don't want to expose the `_editor`.  Do you mean a read-only getter for the editor model?

jasongrout replied:
This was a leftover from a previous implementation that I thought would be convenient.  However, I was only using it once, and even that use wasn't optimal.  So I deleted the getter/setter.

jasongrout commented:
Thanks for the feedback.

jasongrout committed:
Make the signal handler name clearer

Thanks to @sccolbert for the suggestion.

jasongrout committed:
Delete the text getter/setter for the codemirror widget

Interaction with the text should happen through the model.

Also, make internal functions protected.

Thanks to @sccolbert for the suggestion.

blink1073 commented on src/InputAreaWidget.ts:
```diff
@@ -0,0 +1,252 @@
+  /**
+   * Change handler for model updates.
+   */
+  protected _onModelStateChanged(sender: ITextEditorViewModel, args: IChangedArgs<any>) {
```
We've been using a leading `_` for `private` only.

blink1073 replied:
Thoughts?

sccolbert replied:
+1 to leading `_` on `private` only.

jasongrout replied:
Fixed.  Thanks.

blink1073 commented on package.json:
```diff
@@ -1,25 +1,50 @@
+  "devDependencies": {
+    "nodemon": "^1.8.0",
+    "rimraf": "^2.4.3",
+    "steal": "^0.12.3",
```
This repo won't be using steal (also don't need the `system` config in this file).

jasongrout replied:
The example uses steal.  Should we put that in a separate package json in the examples directory?

blink1073 replied:
No, I'll update the example to use SystemJS after we merge this.

blink1073 replied:
Wait, why were you using steal instead of a script tag?

blink1073 replied:
Oh, I see, you're using it instead of browserify.

jasongrout committed:
Delete _ in protected method name.

blink1073 commented:
Are we ready to merge this then?

jasongrout commented:
I rebased in order to remove the experimental js code, instead of moving it to a new directory.  It's a lot cleaner that way, and the code is still in the history if we want to look at it.

blink1073 commented:
LGTM, merge?

jasongrout commented:
I'm ready for it to merge.

We'll need to iterate on the tests.

blink1073 commented:
Yep, iterating, thanks!

jasongrout commented:
And now of course I see that I didn't update the readme nor did I delete the js tests...
